protocols/raft/src/main/java/io/atomix/protocols/raft/partition/RaftPartition.java

```diff
@@ -182,14 +200,21 @@ public RaftPartitionClient getClient() {
    * Creates a Raft server.
    */
   protected RaftPartitionServer createServer(PartitionManagementService managementService) {
-    return new RaftPartitionServer(
+    final RaftPartitionServer raftPartitionServer = new RaftPartitionServer(
         this,
         config,
         managementService.getMembershipService().getLocalMember().id(),
         managementService.getMembershipService(),
         managementService.getMessagingService(),
         managementService.getPrimitiveTypes(),
         threadContextFactory);
+
+    if (!deferredRoleChangeListeners.isEmpty()) {
+      deferredRoleChangeListeners.forEach(raftPartitionServer::addRoleChangeListener);
```

Comment:
@Zelldon after more thought I think it's safer if this operation was moved to after server variable is assigned a value.

Reply:
Thinking about this further I'm wondering if it's too specific to your implementation and it would make more sense exposing a getter for server where by you'd chain a completable future to the Atomix start cf, then iterate the partitions and then getServer().addRole.. then again I guess you may miss the first event fired by join. @kuujo wdyt?